crates/runtime/src/http/v1.rs

```diff
@@ -1178,3 +1178,271 @@ pub(crate) mod embed {
         }
     }
 }
+
+pub(crate) mod assist {
+    use arrow::array::StringArray;
+    use axum::{
+        http::StatusCode,
+        response::{IntoResponse, Response},
+        Extension, Json,
+    };
+    use datafusion::{datasource::TableProvider, sql::TableReference};
+
+    use itertools::Itertools;
+    use serde::{Deserialize, Serialize};
+    use std::{collections::HashMap, sync::Arc};
+    use tokio::sync::RwLock;
+
+    use crate::{
+        datafusion::DataFusion, embeddings::table::EmbeddingTable, EmbeddingModelStore,
+        LLMModelStore,
+    };
+
+    #[derive(Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "lowercase")]
+    pub struct Request {
+        pub text: String,
+
+        /// The model to use for chat completion. The embedding model is determined via the data source (that has the associated embeddings).
+        #[serde(rename = "use", default = "default_model")]
+        pub model: String,
+
+        /// Which datasources in the [`DataFusion`] instance to retrieve data from.
+        #[serde(rename = "from", default)]
+        pub data_source: Vec<String>,
+    }
+
+    fn default_model() -> String {
+        "embed".to_string()
+    }
+
+    async fn create_input_embeddings(
+        input: &str,
+        embeddings_to_run: Vec<String>,
+        embeddings: Arc<RwLock<EmbeddingModelStore>>,
+    ) -> Result<HashMap<String, Vec<f32>>, Box<dyn std::error::Error>> {
+        let mut embedded_inputs: HashMap<String, Vec<f32>> = HashMap::new();
+        for (name, model) in embeddings
+            .read()
+            .await
+            .iter()
+            .filter(|(model_name, _)| embeddings_to_run.contains(model_name))
+        {
+            match model
```

Comment:
a large part of the body should be a method in EmbeddingModelStore

exposing read(), write() here feels like exposing too much thing in this layer

Reply:
See https://github.com/spiceai/spiceai/issues/1617